models/project.go

```diff
@@ -0,0 +1,300 @@
+	Page     int
+	IsClosed util.OptionalBool
+	SortType string
+	Type     ProjectType
+}
+
+// GetProjects returns a list of all projects that have been created in the repository
+func GetProjects(opts ProjectSearchOptions) ([]*Project, error) {
+	return getProjects(x, opts)
+}
+
+func getProjects(e Engine, opts ProjectSearchOptions) ([]*Project, error) {
+
+	projects := make([]*Project, 0, setting.UI.IssuePagingNum)
+
+	sess := e.Where("repo_id = ?", opts.RepoID)
+	switch opts.IsClosed {
+	case util.OptionalBoolTrue:
+		sess = sess.Where("is_closed = ?", true)
+	case util.OptionalBoolFalse:
+		sess = sess.Where("is_closed = ?", false)
+	}
+
+	if opts.Type > 0 {
+		sess = sess.Where("type = ?", opts.Type)
+	}
+
+	if opts.Page > 0 {
+		sess = sess.Limit(setting.UI.IssuePagingNum, (opts.Page-1)*setting.UI.IssuePagingNum)
+	}
+
+	switch opts.SortType {
+	case "oldest":
+		sess.Desc("created_unix")
+	case "recentupdate":
+		sess.Desc("updated_unix")
+	case "leastupdate":
+		sess.Asc("updated_unix")
+	default:
+		sess.Asc("created_unix")
+	}
+
+	return projects, sess.Find(&projects)
+}
+
+// NewProject creates a new Project
+func NewProject(p *Project) error {
+	if !IsProjectBoardTypeValid(p.BoardType) {
+		p.BoardType = None
+	}
+
+	if !IsProjectTypeValid(p.Type) {
+		return errors.New("project type is not valid")
+	}
+
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	if _, err := sess.Insert(p); err != nil {
+		return err
+	}
+
+	if _, err := sess.Exec("UPDATE `repository` SET num_projects = num_projects + 1 WHERE id = ?", p.RepoID); err != nil {
+		return err
+	}
+
+	if err := createBoardsForProjectsType(sess, p); err != nil {
+		return err
+	}
+
+	return sess.Commit()
+}
+
+// GetProjectByID returns the projects in a repository
+func GetProjectByID(id int64) (*Project, error) {
+	return getProjectByID(x, id)
+}
+
+func getProjectByID(e Engine, id int64) (*Project, error) {
+	p := new(Project)
+
+	has, err := e.ID(id).Get(p)
+	if err != nil {
+		return nil, err
+	} else if !has {
+		return nil, ErrProjectNotExist{ID: id}
+	}
+
+	return p, nil
+}
+
+// UpdateProject updates project properties
+func UpdateProject(p *Project) error {
+	return updateProject(x, p)
+}
+
+func updateProject(e Engine, p *Project) error {
+	p.UpdatedUnix = timeutil.TimeStampNow()
+	_, err := e.ID(p.ID).Cols(
+		"title",
+		"description",
+		"updated_unix",
+	).Update(p)
+	return err
+}
+
+func countRepoProjects(e Engine, repoID int64) (int64, error) {
+	return e.Where("repo_id=? AND type=?", repoID, ProjectTypeRepository).
+		Count(new(Project))
+}
+
+func countRepoClosedProjects(e Engine, repoID int64) (int64, error) {
+	return e.
+		Where("repo_id=? AND type=? AND is_closed=?", repoID, ProjectTypeRepository, true).
+		Count(new(Project))
+}
+
+// ChangeProjectStatus toggle a project between opened and closed
+func ChangeProjectStatus(p *Project, isClosed bool) error {
+
+	repo, err := GetRepositoryByID(p.RepoID)
```

Comment:
Why are you getting this outside of the transaction?

Reply:
Actually, why are you getting it at all?

Reply:
good catch - changed it